### PR TITLE
Adding bulk-indexing and improving a bit the name indexing for bleve and elasticsearch

### PR DESCRIPTION
**Pull request by jespino**
#### Summary
Adding bulk-indexing and improving a bit the name indexing for bleve and elasticsearch

#### Jira Ticket
[MM-32874](https://mattermost.atlassian.net/browse/MM-32874)

#### Related PR
Enterpise: mattermost/enterprise#901

```relase-note
NONE
```

### CODE REVIEW

File: services/searchengine/bleveengine/bleve.go [hunk 1]
```diff
@@ -249,6 +249,9 @@ func (b *BleveEngine) deleteIndexes() *model.AppError {
 	if err := os.RemoveAll(b.getIndexDir(ChannelIndex)); err != nil {
 		return model.NewAppError("Bleveengine.PurgeIndexes", "bleveengine.purge_channel_index.error", nil, err.Error(), http.StatusInternalServerError)
 	}
+	if err := os.RemoveAll(b.getIndexDir(FileIndex)); err != nil {
+		return model.NewAppError("Bleveengine.PurgeIndexes", "bleveengine.purge_channel_index.error", nil, err.Error(), http.StatusInternalServerError)
```
Comment (ethervoid) [inline services/searchengine/bleveengine/bleve.go:253]:
```suggestion
		return model.NewAppError("Bleveengine.PurgeIndexes", "bleveengine.purge_file_index.error", nil, err.Error(), http.StatusInternalServerError)
```

File: services/searchengine/bleveengine/common.go [hunk 2]
```diff
@@ -125,6 +125,13 @@ func BLVPostFromPostForIndexing(post *model.PostForIndexing) *BLVPost {
 	}
 }
 
+func splitFilenameWords(name string) string {
```
Comment (ethervoid) [inline services/searchengine/bleveengine/common.go:128]: Given that this function is already in the enterprise code too we can move it to utils and use it in both places

File: store/sqlstore/file_info_store.go [hunk 3]
```diff
@@ -593,3 +593,40 @@ func (fs SqlFileInfoStore) Search(paramsList []*model.SearchParams, userId, team
...
+		Where("DeleteAt = 0")
+
+	queryString, args, err := query.ToSql()
+	if err != nil {
+		return int64(0), errors.Wrap(err, "count_tosql")
+	}
+
+	count, err := fs.GetReplica().SelectInt(queryString, args...)
+	if err != nil {
+		return int64(0), errors.Wrap(err, "failed to count Users")
```
Comment (ethervoid) [inline store/sqlstore/file_info_store.go:610]:
```suggestion
		return int64(0), errors.Wrap(err, "failed to count Files")
```

File: store/sqlstore/file_info_store.go [hunk 4]
```diff
@@ -593,3 +593,40 @@ func (fs SqlFileInfoStore) Search(paramsList []*model.SearchParams, userId, team
...
+		From("FileInfo as fi").
+		LeftJoin("Posts AS p ON fi.PostId = p.Id").
+		Where(sq.GtOrEq{"fi.CreateAt": startTime}).
+		Where(sq.Lt{"fi.CreateAt": endTime}).
+		OrderBy("fi.CreateAt").
+		Limit(uint64(limit)).
+		ToSql()
+	_, err := fs.GetSearchReplica().Select(&files, sql, args...)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to find Users")
```
Comment (ethervoid) [inline store/sqlstore/file_info_store.go:628]:
```suggestion
		return nil, errors.Wrap(err, "failed to find Files")
```

File: store/sqlstore/file_info_store.go [hunk 5]
```diff
@@ -593,3 +593,40 @@ func (fs SqlFileInfoStore) Search(paramsList []*model.SearchParams, userId, team
 	list.MakeNonNil()
 	return list, nil
 }
+
+func (fs SqlFileInfoStore) CountAll() (int64, error) {
```
Comment (ethervoid) [inline store/sqlstore/file_info_store.go:597]: Maybe we should include some tests for these two new methods in the store

**Review by ethervoid — CHANGES_REQUESTED** [review]

File: services/searchengine/bleveengine/common.go (hunk 2 above)
Comment (jespino) [reply services/searchengine/bleveengine/common.go:128]: I prefer to keep some duplication there than moving that to `utils` which means nothing, and is going to add an artificial (and unneeded) coupling with the `utils` package. I'm ok moving this into the searchengine package, but I don't see the need.

File: store/sqlstore/file_info_store.go (hunk 5 above)
Comment (jespino) [reply store/sqlstore/file_info_store.go:597]: done

**Review by ethervoid — APPROVED** [review]
LGTM! Amazing work buddy!

**Comment by jespino** [comment]
@ethervoid @mgdelacroix Sorry for asking the review again, but I added the feature flag to avoid indexing until the feature is fully enabled.

**Comment by jespino** [comment]
@crspeller I added a last commit for adding the feature flag, looks like super light weight, so... I don't know if I'm missing something.

**Review by crspeller — APPROVED** [review]
@jespino That was one of my design goals. Make it really easy to add one 😄 
Looks good!

**Review by ethervoid — APPROVED** [review]
LGTM! Pretty neat how to add it :)

**Review by mgdelacroix — APPROVED** [review]
LGTM! :+1:

**Review by ogi-m — APPROVED** [review]
Tested and passed - reindexing working as expected for ES and Bleve

**Pull request merged**